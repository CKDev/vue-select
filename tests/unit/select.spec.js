import { mount, shallowMount } from '@vue/test-utils'
import Select from '@/select.vue'

import States from './data/states.json'
import Cocktails from './data/cocktails.json'

describe('select.vue', () => {
  it('should render a vue instance', () => {
    const wrapper = shallowMount(Select)
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  it('should have an empty array as default options', () => {
    const wrapper = shallowMount(Select)
    expect(wrapper.props().options).toEqual([])
  })

  it('should not be open by default', () => {
    const wrapper = shallowMount(Select)
    expect(wrapper.vm.open).toBeFalsy()
  })

  it('should not be focused by default', () => {
    const wrapper = shallowMount(Select)
    expect(wrapper.vm.focused).toBeFalsy()
  })

  it('should not be focused by default', () => {
    const wrapper = shallowMount(Select)
    expect(wrapper.vm.focused).toBeFalsy()
  })

  it('should display the placeholder if no value selected', () => {
    const wrapper = shallowMount(Select, { propsData: { placeholder: 'Hello World' } })
    expect(wrapper.find('.label').text()).toEqual('Hello World')
  })

  it('should be opened when label is clicked', () => {
    const wrapper = shallowMount(Select)
    wrapper.find('.label').trigger('click')
    expect(wrapper.vm.open).toBeTruthy()
    expect(wrapper).toMatchSnapshot()
  })

  it('should change the selection to the clicked option', () => {
    const wrapper = mount(Select, { propsData: { options: States } })
    const options = wrapper.findAll('.v-option:not(.is-disabled)')
    const option = options.at(Math.floor(Math.random() * options.length))
    expect(wrapper.vm.selection).toBeFalsy()
    option.trigger('click')
    expect(wrapper.vm.selection).toEqual(option.vm.option)
  })

  it('should have a single hovered option on mouseover', () => {
    const wrapper = mount(Select, { propsData: { options: States } })
    const options = wrapper.findAll('.v-option:not(.is-disabled)')
    const option = options.at(Math.floor(Math.random() * options.length))
    expect(options.filter(o => o.vm.option.state.hovered).length).toEqual(0)
    option.trigger('mouseover')
    expect(options.filter(o => o.vm.option.state.hovered).length).toEqual(1)
  })

  it('should emit the input event when the selection changed', (done) => {
    const wrapper = mount(Select, { propsData: { options: States } })
    const options = wrapper.findAll('.v-option:not(.is-disabled)')
    const option = options.at(Math.floor(Math.random() * options.length))
    option.trigger('click')
    wrapper.vm.$nextTick(() => {
      expect(wrapper.emitted().input).toBeTruthy()
      done()
    })
  })

  it('should emit the change event when the selection changed', (done) => {
    const wrapper = mount(Select, { propsData: { options: States } })
    const options = wrapper.findAll('.v-option:not(.is-disabled)')
    const option = options.at(Math.floor(Math.random() * options.length))
    option.trigger('click')
    wrapper.vm.$nextTick(() => {
      expect(wrapper.emitted().change).toBeTruthy()
      done()
    })
  })

  it('should move to prev/next available option on arrow key press', (done) => {
    const wrapper = mount(Select, { propsData: { options: States } })
    const options = wrapper.findAll('.v-option:not(.is-disabled)')
    const initialIndex = options.at(Math.floor(options.length / 2)).vm.option.state.index
    wrapper.setData({ open: true, highlightIndex: initialIndex })
    wrapper.trigger('keydown.down')
    console.log(wrapper.find('.v-option.is-highlighted').vm.option)

    wrapper.vm.$nextTick(() => {
      expect(wrapper.find('.v-option.is-highlighted').vm.option.state.index).toBeGreaterThan(initialIndex)
      wrapper.trigger('keydown.up')
      wrapper.trigger('keydown.up')
      expect(wrapper.find('.v-option.is-highlighted').vm.option.state.index).toBeLessThan(initialIndex)
      done()
    })
  })

  it('should highlight the currently hovered option on keypress', () => {
    const wrapper = mount(Select, { propsData: { options: States } })
    const options = wrapper.findAll('.v-option:not(.is-disabled)')
    const initialIndex = options.at(Math.floor(options.length / 2)).vm.option.state.index
    wrapper.setData({ open: true, hoverIndex: initialIndex })
    expect(wrapper.vm.highlightIndex).toEqual(-1)
    wrapper.trigger('keydown.down')
    expect(wrapper.vm.highlightIndex).toEqual(initialIndex)
  })

  it('should highlight the first available option on keypress', () => {
    const wrapper = mount(Select, { propsData: { options: States } })
    const option = wrapper.find('.v-option:not(.is-disabled)')
    wrapper.trigger('keydown.down')
    expect(wrapper.vm.highlightIndex).toEqual(option.vm.option.state.index)
    expect(wrapper).toMatchSnapshot()
  })

  describe('when closed', () => {
    it('should open when space key is pressed', () => {
      const wrapper = shallowMount(Select)
      wrapper.setData({ focused: true })
      wrapper.trigger('keydown.space')
      expect(wrapper.vm.open).toBeTruthy()
    })
  })

  describe('when open', () => {
    it('should close when the escape key is pressed', () => {
      const wrapper = shallowMount(Select)
      wrapper.setData({ open: true })
      wrapper.trigger('keydown.esc')
      expect(wrapper.vm.open).toBeFalsy()
    })

    it('should close when the tab key is pressed', () => {
      const wrapper = shallowMount(Select)
      wrapper.setData({ open: true, focused: true })
      wrapper.trigger('keydown.tab')
      expect(wrapper.vm.open).toBeFalsy()
    })

    it('should close when the enter key is pressed', () => {
      const wrapper = shallowMount(Select)
      wrapper.setData({ open: true })
      wrapper.trigger('keydown.enter')
      expect(wrapper.vm.open).toBeFalsy()
    })

    it('should have an open state when opened', () => {
      const wrapper = shallowMount(Select)
      wrapper.find('.label').trigger('click')
      expect(wrapper.vm.open).toBeTruthy()
    })

    it('should have a focused state when focused', () => {
      const wrapper = shallowMount(Select)
      wrapper.trigger('focus')
      expect(wrapper.vm.focused).toBeTruthy()
    })

    it('should close when no longer focused', () => {
      const wrapper = shallowMount(Select)
      wrapper.trigger('focus')
      wrapper.setData({ open: true })
      wrapper.trigger('blur')
      expect(wrapper.vm.open).toBeFalsy()
    })
  })

  describe('with no highlighted options', () => {
    it('should select the hovered option when tab key pressed', () => {
      const wrapper = mount(Select, { propsData: { options: States } })
      const options = wrapper.findAll('.v-option:not(.is-disabled)')
      const option = options.at(Math.floor(Math.random() * options.length))
      wrapper.setData({ open: true, focused: true, hoverIndex: option.vm.option.state.index })
      wrapper.trigger('keydown.tab')
      expect(wrapper.vm.selection).toEqual(option.vm.option)
    })
  })

  describe('with a highlighted option', () => {
    it('should select the highlighted option when tab key pressed', () => {
      const wrapper = mount(Select, { propsData: { options: States } })
      const options = wrapper.findAll('.v-option:not(.is-disabled)')
      const option = options.at(Math.floor(Math.random() * options.length))
      wrapper.setData({ open: true, focused: true, highlightIndex: option.vm.option.state.index })
      wrapper.trigger('keydown.tab')
      expect(wrapper.vm.selection).toEqual(option.vm.option)
    })
  })
})
