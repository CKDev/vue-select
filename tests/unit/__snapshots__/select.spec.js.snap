// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`select.vue should be opened when label is clicked 1`] = `
<div role="listbox" tabindex="0" class="v-select is-open is-placeholding is-down">
  <div class="label"></div>
  <div class="options" style=""></div> <select tabindex="-1" class="select"></select>
</div>
`;

exports[`select.vue should highlight the first available option on keypress 1`] = `
<div role="listbox" tabindex="0" class="v-select is-down">
  <div class="label">Arizona</div>
  <div aria-hidden="true" class="options" style="animation-duration: 0s;">
    <div role="option" aria-disabled="true" aria-selected="false" class="v-option is-disabled" style="padding-left: 12px;">
      <div class="label">Alabama</div>
    </div>
    <div role="option" aria-disabled="true" aria-selected="false" class="v-option is-disabled" style="padding-left: 12px;">
      <div class="label">Alaska</div>
    </div>
    <div role="option" aria-selected="true" class="v-option is-selected is-highlighted is-hovered" style="padding-left: 12px;">
      <div class="label">Arizona</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">Arkansas</div>
    </div>
    <div role="option" aria-disabled="true" aria-selected="false" class="v-option is-disabled" style="padding-left: 12px;">
      <div class="label">California</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label"><strong>Colorado</strong></div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">Connecticut</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">Delaware</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">District Of Columbia</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">Florida</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">Georgia</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">Hawaii</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">Idaho</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">Illinois</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">Indiana</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">Iowa</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">Kansas</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">Kentucky</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">Louisiana</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">Maine</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">Maryland</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">Massachusetts</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">Michigan</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">Minnesota</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">Mississippi</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">Missouri</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">Montana</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">Nebraska</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">Nevada</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">New Hampshire</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">New Jersey</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">New Mexico</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">New York</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">North Carolina</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">North Dakota</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">Ohio</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">Oklahoma</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">Oregon</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">Pennsylvania</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">Rhode Island</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">South Carolina</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">South Dakota</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">Tennessee</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">Texas</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">Utah</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">Vermont</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">Virginia</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">Washington</div>
    </div>
    <div role="option" aria-selected="false" class="v-option" style="padding-left: 12px;">
      <div class="label">West Virginia</div>
    </div>
    <div role="option" aria-disabled="true" aria-selected="false" class="v-option is-disabled" style="padding-left: 12px;">
      <div class="label">Wisconsin</div>
    </div>
    <div role="option" aria-disabled="true" aria-selected="false" class="v-option is-disabled" style="padding-left: 12px;">
      <div class="label">Wyoming</div>
    </div>
  </div> <select tabindex="-1" class="select">
    <option value="AL" disabled="">Alabama</option>
    <option value="AK" disabled="">Alaska</option>
    <option value="AZ" selected="">Arizona</option>
    <option value="AR">Arkansas</option>
    <option value="CA" disabled="">California</option>
    <option value="CO">Colorado</option>
    <option value="CT">Connecticut</option>
    <option value="DE">Delaware</option>
    <option value="DC">District Of Columbia</option>
    <option value="FL">Florida</option>
    <option value="GA">Georgia</option>
    <option value="HI">Hawaii</option>
    <option value="ID">Idaho</option>
    <option value="IL">Illinois</option>
    <option value="IN">Indiana</option>
    <option value="IA">Iowa</option>
    <option value="KS">Kansas</option>
    <option value="KY">Kentucky</option>
    <option value="LA">Louisiana</option>
    <option value="ME">Maine</option>
    <option value="MD">Maryland</option>
    <option value="MA">Massachusetts</option>
    <option value="MI">Michigan</option>
    <option value="MN">Minnesota</option>
    <option value="MS">Mississippi</option>
    <option value="MO">Missouri</option>
    <option value="MT">Montana</option>
    <option value="NE">Nebraska</option>
    <option value="NV">Nevada</option>
    <option value="NH">New Hampshire</option>
    <option value="NJ">New Jersey</option>
    <option value="NM">New Mexico</option>
    <option value="NY">New York</option>
    <option value="NC">North Carolina</option>
    <option value="ND">North Dakota</option>
    <option value="OH">Ohio</option>
    <option value="OK">Oklahoma</option>
    <option value="OR">Oregon</option>
    <option value="PA">Pennsylvania</option>
    <option value="RI">Rhode Island</option>
    <option value="SC">South Carolina</option>
    <option value="SD">South Dakota</option>
    <option value="TN">Tennessee</option>
    <option value="TX">Texas</option>
    <option value="UT">Utah</option>
    <option value="VT">Vermont</option>
    <option value="VA">Virginia</option>
    <option value="WA">Washington</option>
    <option value="WV">West Virginia</option>
    <option value="WI" disabled="">Wisconsin</option>
    <option value="WY" disabled="">Wyoming</option>
  </select>
</div>
`;
